src/esw/drive/odrive_control.py

```diff
@@ -0,0 +1,656 @@
+    right: float
+
+    def __init__(self, left: float = 0.0, right: float = 0.0) -> None:
+        self.left = left
+        self.right = right
+
+    def __eq__(self, other):
+        """Overrides the equality comparator operator."""
+        return self.left == other.left and self.right == other.right
+
+
+class DisconnectedError(Exception):
+    """An exception thrown when the Jetson is unable to talk with the ODrive."""
+
+    pass
+
+
+class ODriveEvent(Enum):
+    """Dictates the behavior of the changing of States.
+    The ODriveEvents reveal the following: When the ODrive is disconnected,
+    is trying to arm, or has encountered an error.
+    """
+
+    DISCONNECTED_ODRIVE_EVENT = 0
+    ARM_CMD_EVENT = 1
+    ODRIVE_ERROR_EVENT = 2
+
+
+class Modrive:
+    """Abstracts the behavior of the ODrive.
+    The ODriveBridge creates the same Modrive object
+    every time it tries to connect to an ODrive. The Modrive object
+    has functions that allows it to get encoder data, get state data,
+    set velocity, and much more.
+    :param _axes: A dictionary that maps left or right to an ODrive axes
+        object.
+    :param _meters_to_turns_ratio_by_side: A dictionary that maps left or
+        right to a multiplier that can be used to convert turns into meters.
+    :param _odrive: An ODrive object.
+    :param _turns_to_raw_ratio_by_side: A dictionary that maps left or right
+        to the vel cmd multiplier that can be used to convert a raw [0, 1]
+        command to desired turns.
+    :param _usb_lock: A lock that prevents multiple threads from accessing
+        ODrive objects simultaneously.
+    :param _watchdog_timeout: A float that represents the watchdog timeout.
+    """
+
+    _axes: Dict[str, Any]
+    _meters_to_turns_ratio_by_side: Dict[str, float]
+    _odrive: Any
+    _turns_to_raw_ratio_by_side: Dict[str, float]
+    _usb_lock: threading.Lock
+    _watchdog_timeout: float
+
+    def __init__(self, odr) -> None:
+        self._odrive = odr
+        _side_by_axis: Dict[int, str] = {
+            rospy.get_param("/odrive/axis/left"): "left",
+            rospy.get_param("/odrive/axis/right"): "right",
+        }
+
+        self._axes = {_side_by_axis[0]: self._odrive.axis0, _side_by_axis[1]: self._odrive.axis1}
+        self._meters_to_turns_ratio_by_side = {
+            "left": rospy.get_param("/odrive/ratio/meters_to_turns_ratio_left"),
+            "right": rospy.get_param("/odrive/ratio/meters_to_turns_ratio_right"),
+        }
+        self._turns_to_raw_ratio_by_side = {
+            "left": rospy.get_param("/odrive/ratio/turns_to_raw_ratio_left"),
+            "right": rospy.get_param("/odrive/ratio/turns_to_raw_ratio_right"),
+        }
+        self._usb_lock = threading.Lock()
+        self._watchdog_timeout = rospy.get_param("/odrive/config/watchdog_timeout")
+
+    def __getattr__(self, attr: Any) -> Any:
+        """
+        TODO - I'M BAD AT PYTHON, SOMEONE ADD A COMMENT WHY THIS IS HERE.
```

Comment:
Can we clear this up before merging? I'm not really familiar with attributes in python but seems to flag a condition where the Modrive object is None

Reply:
ya i forgot to resolve this todo... ill go add a comment if ur explanation is correct

Reply:
Fixed